// quiz-app/src/components/Quiz.js
import React, { useState, useEffect } from "react";
import { useLocation, useNavigate } from "react-router-dom";
import { Container, Typography, Button, Card, CardContent, LinearProgress, Box } from "@mui/material";
import { motion } from "framer-motion";

const useQuery = () => {
  return new URLSearchParams(useLocation().search);
};

const Quiz = () => {
  const query = useQuery();
  const navigate = useNavigate();
  
  const numQuestions = parseInt(query.get("num")) || 10;
  const mode = query.get("mode") || "chill";
  const timeLimit = parseInt(query.get("time")) || 30;
  
  const [questions, setQuestions] = useState([]);
  const [currentIndex, setCurrentIndex] = useState(0);
  const [selectedAnswer, setSelectedAnswer] = useState(null);
  const [correctAnswer, setCorrectAnswer] = useState(null);
  const [isCorrect, setIsCorrect] = useState(null);
  const [timeLeft, setTimeLeft] = useState(timeLimit * 60);
  const [correctCount, setCorrectCount] = useState(0);
  const [showFeedback, setShowFeedback] = useState(false); // æŽ§åˆ¶åŠ¨ç”»

  useEffect(() => {
    fetch(`http://127.0.0.1:8000/questions?limit=${numQuestions}`)
      .then((res) => res.json())
      .then((data) => {
        console.log("ðŸ” API è¿”å›žæ•°æ®:", data);
        if (data.questions && data.questions.length > 0) {
          setQuestions(data.questions);
        } else {
          console.error("æ²¡æœ‰æ”¶åˆ°é¢˜ç›®æ•°æ®:", data);
        }
      })
      .catch((err) => console.error("ðŸš¨ åŠ è½½é¢˜ç›®å¤±è´¥:", err));
  }, [numQuestions]);

  useEffect(() => {
    if (mode === "timed" && timeLeft > 0) {
      const timer = setInterval(() => {
        setTimeLeft((prev) => prev - 1);
      }, 1000);
      return () => clearInterval(timer);
    }
  }, [mode, timeLeft]);

  useEffect(() => {
    if (mode === "timed" && timeLeft === 0) {
      nextQuestion();
    }
  }, [mode, timeLeft]);

  const handleAnswer = (answerKey) => {
    setSelectedAnswer(answerKey);  // åªå­˜ "A"ã€"B"ã€"C"ã€"D"
  };

  const nextQuestion = () => {
    if (!selectedAnswer) return;

    const currentQuestion = questions[currentIndex];

    fetch("http://127.0.0.1:8000/check_answer", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ question_number: currentQuestion.question_number, user_answer: selectedAnswer }),
    })
      .then((res) => res.json())
      .then((data) => {
        console.log("ðŸ” API è¿”å›ž:", data);
        console.log("âœ… é¢˜ç›®:", currentQuestion.question_content);
        console.log("ðŸ’¡ ç”¨æˆ·é€‰æ‹©:", selectedAnswer);
        console.log("ðŸŽ¯ æ­£ç¡®ç­”æ¡ˆ:", currentQuestion.correct_answer);

        const correct = Boolean(data.is_correct);
        setIsCorrect(correct);
        setCorrectAnswer(currentQuestion.correct_answer);
        setShowFeedback(true);  // è¿™é‡Œå¼€å¯åŠ¨ç”»

        if (correct) {
          setCorrectCount((prev) => prev + 1);
        }

        setTimeout(() => {
          if (currentIndex + 1 < questions.length) {
            setCurrentIndex(currentIndex + 1);
            setSelectedAnswer(null);
            setIsCorrect(null);
            setCorrectAnswer(null);
            setShowFeedback(false);
            if (mode === "timed") setTimeLeft(timeLimit * 60);
          } else {
            navigate(`/results?total=${questions.length}&correct=${correctCount}`);
          }
        }, 1000);
      })
      .catch(err => console.error("ðŸš¨ æ£€æŸ¥ç­”æ¡ˆè¯·æ±‚å¤±è´¥:", err));
  };

  if (questions.length === 0) return <Typography>åŠ è½½ä¸­...</Typography>;

  return (
    <Container maxWidth="sm" style={{ backgroundColor: "#2C2C2C", minHeight: "100vh", textAlign: "center", padding: "20px" }}>
      <Typography variant="h5" style={{ color: "white", fontWeight: "bold", marginBottom: "10px" }}>
        é¢˜ç›® {currentIndex + 1} / {questions.length}
      </Typography>

      <Card variant="outlined" style={{ padding: "20px", backgroundColor: "#444", color: "white", marginBottom: "20px" }}>
        <CardContent>
          <Typography variant="h6">{questions[currentIndex].question_content}</Typography>
        </CardContent>
      </Card>

      {Object.entries(questions[currentIndex].options).map(([key, option]) => (
        <motion.div 
          key={key}
          animate={showFeedback && selectedAnswer === key ? 
            (isCorrect ? { scale: [1, 1.1, 1] } : { x: [-5, 5, -5, 5, 0] }) // æ­£ç¡®ç­”æ¡ˆæ”¾å¤§ï¼Œé”™è¯¯ç­”æ¡ˆæŠ–åŠ¨
            : {}}
          transition={{ duration: 0.3 }}
        >
          <Button
            variant="contained"
            fullWidth
            onClick={() => handleAnswer(key)}
            sx={{
              marginTop: "10px",
              fontSize: "18px",
              backgroundColor:
                showFeedback
                  ? key === correctAnswer
                    ? "#4CAF50"  // é€‰å¯¹äº†å˜ç»¿è‰²
                    : key === selectedAnswer
                    ? "#FF4C4C"  // é€‰é”™äº†å˜çº¢
                    : "#666"
                  : selectedAnswer === key
                  ? "#FFA500"  // é€‰ä¸­ä½†æœªç¡®è®¤ï¼Œæ©™è‰²
                  : "#666",
              "&:hover": { backgroundColor: "#FF8C00" },
              color: "white",
              padding: "12px",
              borderRadius: "12px",
            }}
          >
            {option}
          </Button>
        </motion.div>
      ))}

      {selectedAnswer && (
        <Button variant="contained" color="secondary" onClick={nextQuestion} sx={{ marginTop: "20px" }}>
          ä¸‹ä¸€é¢˜
        </Button>
      )}
    </Container>
  );
};

export default Quiz;
